src/app/CommandHandler.h

```diff
@@ -350,9 +399,21 @@ class CommandHandler
      * @param [in] aData the data for the response.
      */
     template <typename CommandData>
-    void AddResponse(const ConcreteCommandPath & aRequestCommandPath, const CommandData & aData)
+    inline void AddResponse(const ConcreteCommandPath & aRequestCommandPath, const CommandData & aData)
     {
-        if (AddResponseData(aRequestCommandPath, aData) != CHIP_NO_ERROR)
+        DataModelEncoderToTLV<CommandData> encoder(aData);
+        return AddResponseViaEncoder(aRequestCommandPath, CommandData::GetCommandId(), encoder);
+    }
+
+    /**
+     * API for adding a response with a given encoder of TLV data.
+     *
+     * The encoder would generally encode a ClusterName::Commands::CommandName::Type with
+     * the corresponding `GetCommandId` call.
+     */
+    void AddResponseViaEncoder(const ConcreteCommandPath & aRequestCommandPath, CommandId commandId, EncoderToTLV & encoder)
```

Comment:
Again, why the "ViaEncoder" suffix?

Reply:
I had it without `ViaEncoder` at first however then `<template typename T>void AddResponse(const ConcreteCommandPath&, T>` and `void AddResponse(const ConcreteCommandPath&, EncoderToTLV &encoder)` seemed ambigous.

In the mean time I had to add commandId, so I could rename them as well I guess. This was iterated a few times, can be more.